pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/impl/BlockAwareSegmentInputStreamImpl.java

```diff
@@ -122,14 +122,15 @@ private int readEntries() throws IOException {
             Iterator<LedgerEntry> iterator = ledgerEntriesOnce.iterator();
             while (iterator.hasNext()) {
                 LedgerEntry entry = iterator.next();
-                int entryLength = (int) entry.getLength();
+                ByteBuf buf = entry.getEntryBuffer().retain();
```

Comment:
Shouldn't this retain be compensated with a `release()`. I assume the entries returned by this method will be released but the entryBuffer itself would have to be released.

Reply:
The refCount for this retain has its ownership transferred to the composite bytebuf created below, so it will get released when that bytebuf is released.

Reply:
Yes, this is having the same net sum as before this change. My concern was around the `LedgerEntry` themselves. When we call `getEntryBuffer()` the entries are not automatically released. Eg: if we were to remove the `CompositeByteBuf` we still would have to do a release on the entry.

Reply:
they are released when LedgerEntries#close gets called. It gets called through the try-with-resource block. CompositeByteBuf doesn't retain, so we have to call retain explicitly before passing it.

Reply:
Got it, I was missing the `LedgerEntries.close()`